tests/performance/mu-plugins/server-timing.php

```diff
@@ -0,0 +1,43 @@
+<?php
+
+add_action(
+	'template_redirect',
+	function() {
+
+		global $timestart;
+
+		$server_timing_values = array();
+		$template_start       = microtime( true );
+
+		$server_timing_values['before-template'] = $template_start - $timestart;
+
+		ob_start();
+
+		add_action(
+			'shutdown',
+			function() use ( $server_timing_values, $template_start ) {
+
+				global $timestart;
+
+				$output = ob_get_clean();
+
+				$server_timing_values['template'] = microtime( true ) - $template_start;
+
+				$server_timing_values['total'] = $server_timing_values['before-template'] + $server_timing_values['template'];
+
+				$header_values = array();
+				foreach ( $server_timing_values as $slug => $value ) {
+					if ( is_float( $value ) ) {
+						$value = round( $value * 1000.0, 2 );
+					}
+					$header_values[] = sprintf( 'wp-%1$s;dur=%2$s', $slug, $value );
+				}
+				header( 'Server-Timing: ' . implode( ', ', $header_values ) );
+
+				echo $output;
+			},
+			-9999
```

Comment:
We may want to set this to `PHP_INT_MIN`. Since WordPress core polyfills that constant, we can safely use it here. The `-9999` is a bit of a hack only needed if that constant doesn't exist.

Reply:
Isn't it the other way around anyway, the outer hook should be called as early as possible (`PHP_INT_MIN`) and the shutdown hook should run as late as possible (`PHP_INT_MAX`) so that the timing is most accurate?

Reply:
@swissspidy The outer hook should definitely be `PHP_INT_MAX`, since everything that happens in `template_redirect` should be attributed to the first metric still.

For the inner hook, I think we should measure as soon as possible, since at that point the full page has been output, and that's what we care about for the `wp-template` metric. I'm not sure there's anything worth measuring that would happen on `shutdown` anyway.

Reply:
These have been updated to use the relevant PHP constants. I'm going to resolve this, but feel free to reopen if further discussion in necessary.